linalg/constraints.hpp

```diff
@@ -0,0 +1,404 @@
+   /// This applies \f$ -B_s^{-1} B_p \f$.
+   void Eliminate(const Vector& in, Vector& out) const;
+
+   /// Transpose of Eliminate(), applies \f$ -B_p^T B_s^{-T} \f$
+   void EliminateTranspose(const Vector& in, Vector& out) const;
+
+   /// Maps Lagrange multipliers to secondary displacements,
+   /// applies \f$ B_s^{-1} \f$
+   void LagrangeSecondary(const Vector& in, Vector& out) const;
+
+   /// Transpose of LagrangeSecondary()
+   void LagrangeSecondaryTranspose(const Vector& in, Vector& out) const;
+
+   /// Return \f$ -B_s^{-1} B_p \f$ explicitly assembled in mat
+   void ExplicitAssembly(DenseMatrix& mat) const;
+
+private:
+   Array<int> lagrange_tdofs;
+   Array<int> primary_tdofs;
+   Array<int> secondary_tdofs;
+
+   DenseMatrix Bp;
+   DenseMatrix Bs;  // gets inverted in place
+   LUFactors Bsinverse;
+   DenseMatrix BsT;   // gets inverted in place
+   LUFactors BsTinverse;
+   Array<int> ipiv;
+   Array<int> ipivT;
+};
+
+
+/** Collects action of several Eliminator objects to perform elimination of
+    constraints.
+
+    Works in parallel, but each Eliminator must be processor local, and must
+    operate on disjoint degrees of freedom (ie, the primary and secondary dofs
+    for one Eliminator must have no overlap with any dofs from a different
+    Eliminator). */
+class EliminationProjection : public Operator
+{
+public:
+   EliminationProjection(const Operator& A, Array<Eliminator*>& eliminators);
+
+   void Mult(const Vector& x, Vector& y) const override;
+
+   void MultTranspose(const Vector& x, Vector& y) const override;
+
+   /** @brief Assemble this projector as a (processor-local) SparseMatrix.
+
+       Some day we may also want to try approximate variants. */
+   SparseMatrix * AssembleExact() const;
+
+   /** Given Lagrange multiplier right-hand-side \f$ g \f$, return
+       \f$ \tilde{g} \f$ */
+   void BuildGTilde(const Vector& g, Vector& gtilde) const;
+
+   /** After a solve, recover the Lagrange multiplier. */
+   void RecoverMultiplier(const Vector& disprhs,
+                          const Vector& disp, Vector& lm) const;
+
+private:
+   const Operator& Aop;
+   Array<Eliminator*> eliminators;
+};
+
+
+#ifdef MFEM_USE_MPI
+/** @brief Solve constrained system by eliminating the constraint; see
+    ConstrainedSolver
+
+    Solves the system with the operator \f$ P^T A P + Z_P \f$, where P is
+    EliminationProjection and Z_P is the identity on the eliminated dofs. */
+class EliminationSolver : public ConstrainedSolver
+{
+public:
+   /** @brief Constructor, with explicit splitting into primary/secondary dofs.
+
+       This constructor uses a single elimination block (per processor), which
+       provides the most general algorithm but is also not scalable
+
+       The secondary_dofs are eliminated from the system in this algorithm,
+       as they can be written in terms of the primary_dofs. */
+   EliminationSolver(HypreParMatrix& A, SparseMatrix& B,
+                     Array<int>& primary_dofs,
```

Comment:
I assume these are local ordering given the local nature of this operator (and the fact they're `int`s vs `HYPRE_Int`s, but that should be explicitly stated.

Reply:
It it is in local truedof numbering, which is now documented.